### PR TITLE
Add JWE encryption

### PR DESCRIPTION
**Pull request by ThibaudDauce**

### CODE REVIEW

**Review by Manuthor — APPROVED** [review]

File: crate/server/src/config/jwe.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+use josekit::jwk::Jwk as JoseJwk;
+
+#[derive(Debug, Clone)]
+pub struct Jwk(JoseJwk);
+
+impl FromStr for Jwk {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Jwk(JoseJwk::from_reader(&mut s.as_bytes()).unwrap()))
```
[the remark this answers is not in the document]
Comment (ThibaudDauce) [reply crate/server/src/config/jwe.rs]: Done in 0cc7e22

File: crate/client/src/lib.rs [hunk 2]
```diff
@@ -550,12 +561,35 @@ impl KmsRestClient {
...
+            let payload = serialize_compact(
+                serde_json::to_string(&ttlv)
+                    .map_err(|_| {
+                        KmsClientError::UnexpectedError("Cannot transform TTLV to JSON".to_string())
+                    })?
+                    .as_bytes(),
+                &header,
+                &encrypter,
+            )
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (ThibaudDauce) [reply crate/client/src/lib.rs]: 0cc7e22

File: crate/server/src/routes.rs [hunk 3]
```diff
@@ -63,78 +64,93 @@ impl actix_web::error::ResponseError for KmsError {
...
+                let payload = deserialize_compact(&body, &decrypter).unwrap();
+
+                serde_json::from_slice::<TTLV>(&payload.0).unwrap()
+            } else {
+                return Err(KmsError::NotSupported(
+                    "this server doesn't support JWE encryption".to_string(),
+                ))
+            }
+        }
+    };
```
[the remark this answers is not in the document]
Comment (ThibaudDauce) [reply crate/server/src/routes.rs]: Done in d74e7b5

File: crate/cli/Cargo.toml [hunk 4]
```diff
@@ -22,6 +22,7 @@ cosmian_kmip = { path = "../kmip" }
 cosmian_kms_client = { path = "../client" }
 cosmian_kms_utils = { path = "../utils" }
 hex = { workspace = true }
+josekit = "0.8.3"
```
[the remark this answers is not in the document]
Comment (ThibaudDauce) [reply crate/cli/Cargo.toml]: Done in cfa2298

File: crate/client/src/lib.rs [hunk 5]
```diff
@@ -550,12 +561,35 @@ impl KmsRestClient {
...
+
+        request = if let Some(jwe_public_key) = &self.jwe_public_key {
+            let mut header = JweHeader::new();
+            header.set_algorithm("ECDH-ES");
+            header.set_content_encryption("A256GCM");
+            header.set_key_id(jwe_public_key.key_id().unwrap());
+
+            let encrypter = EcdhEsJweAlgorithm::EcdhEs
+                .encrypter_from_jwk(jwe_public_key)
+                .unwrap();
```
[the remark this answers is not in the document]
Comment (ThibaudDauce) [reply crate/client/src/lib.rs]: Done in 0cc7e22

**Review by tbrezot — APPROVED** [review]

**Pull request merged**